### ISSUE TITLE
Debugging possible memory leak?

### ISSUE THREAD

**Issue by stoically** [jellyfin/jellyfin-webos#81]
One of my media files sometimes leads to a webos message popup saying "restarting app to free memory". Doesn't happen with other media files. I'd be happy to provide debug logs but not sure how to?

**Comment by Danacus** [comment]
I had the same issue when viewing photos. Never had any problems with videos.

**Comment by dab2020** [comment]
I hardly doubt that this is a issue on Jellyfin's side, it's most likely the hardware hardware . I mean these TV's are pretty low spec and have a tiny amount of RAM.

How old is your TV and what version of webOS are you running?.

**Comment by stoically** [comment]
Model: LG 49UN71006LB
Version: 5.3.1-29

The fact that other media files larger in size work without this issue was an indicator for me that it might be an app issue. However, havent tested since the 10.8 server release - and using the app from the store might help as well, since running in dev mode seems to put additional stress on the hardware.

**Comment by dab2020** [comment]
[This Guy](https://ddscentral.org/2018/01/unusual-linux-devices-hacking-tvs-part-4-instructions/) found that video's played via the native video player bypass webOS directly:

> All decoding is done through hardware and the decoded video is displayed directly on the screen through decoder’s own internal video output, bypassing webOS entirely.

Try playing a live stream within your browser and see if you get the same error.

**Comment by stoically** [comment]
Thanks for the hint, playing the same file via the browser works fine, no crashes. 👍

Also tried with the app again and server 10.8.0-beta2 (because of https://github.com/jellyfin/jellyfin/issues/8043), but crashed with the low memory error after like 10 minutes again.

**Comment by dab2020** [comment]
Hmm..
Maybe try using 10.8.1 hotfix.

This seems to be a niche issue.

**Comment by stoically** [comment]
As described in the issue, I did test with 10.8.1

**Comment by stoically** [comment]
> This seems to be a niche issue.

Yeah, memory leaks can be sneaky. That's why I initially asked for instructions to debug them myself and/or provide logs.

**Comment by LordDeimos666** [comment]
Happening also to me, while playing big 4k video there's no problem at all.
While playing some small anime episode with ASS subtitles enabled, disabling ASS subtitles does not give the memory leak.

**Comment by Arzte** [comment]
I can confirm with @LordDeimos666's issue of subtitles causing issues. It's fine when there isn't a lot of text on screen, or not a lot of text in quick secession. (EG sign/songs subtitles available for some anime) However, outside of that, I've had issues with webOS killing the client if subtitles come in too rapidly. 

I do wonder if #107 might be relevant because it doesn't seem to be killed for memory until a couple of seconds after whichever set of subtitles shows on the screen, usually out of sync to a degree.

Playing from the browser, or forcing the subtitles to be burned in do not have the same issue.

**Comment by maximushugus** [comment]
I can also confirm, on my rooted LG TV on WebOS 05.40.26, if I play a movie without enabling subtitles there is no problem. As soon as I enable ASS subtitles, the app crashes in less than a second. I can SSH on the TV, and I can see the memory leak, with used RAM going up until the crash. There is no problem with other types of subtitle.
How could we debug this problem ?

**Comment by Akiskaloz** [comment]
" if I play a movie without enabling subtitles there is no problem. As soon as I enable ASS subtitles, the app crashes in less than a second." 
I have the same issue, it doesn't matter what kind of movie as soon as I enable subtitles it goes through the "App will restarts to save up memory"

**Comment by Pedro-MC** [comment]
My TV (LG WebOS UM7000PLA v5.40.61) has the exact same issue, as described above, and using srt subtitles instead of ass solves the issue.

**Comment by krullmizter** [comment]
Same problem with my setup as well. 4GB RAM on my Jellyfin instance, and an old LG TV. With .ass subtitles I get the "restarting app to free memory" crash. Without .ass or with .srt subtitles there is no problem.

**Comment by ngdangtu-vn** [comment]
I have the same problem from ASS subtitles too:

```txt
video_title
Container: mkv
Size: 677.3 MiB

Video
Title: 1080p H264 SDR
Codec: H264
AVC: Yes
Profile: High
Level: 40
Resolution: 1920x1080
Aspect ratio: 16:9
Anamorphic: No
Interlaced: No
Framerate: 23.976025
Bitrate: 3996 kbps
Bit depth: 8 bit
Video range: SDR
Video range type: SDR
Pixel format: yuv420p
Ref frames: 1
NAL: 4

Audio
Title: AAC - Stereo - Default
Codec: AAC
AVC: No
Profile: LC
Layout: stereo
Channels: 2 ch
Bitrate: 128 kbps
Sample rate: 44100 Hz
Default: Yes
Forced: No
External: No

Subtitle
Title: English(US) - Undefined - Default - ASS
Codec: ASS
AVC: No
Default: Yes
Forced: No
External: No

Subtitle
Title: Arabic - Undefined - ASS
Codec: ASS
AVC: No
Default: No
Forced: No
External: No

Subtitle
Title: Deutsch - Undefined - ASS
Codec: ASS
AVC: No
Default: No
Forced: No
External: No

Subtitle
Title: Espaأ±ol(Espaأ±a) - Undefined - ASS
Codec: ASS
AVC: No
Default: No
Forced: No
External: No

Subtitle
Title: Espaأ±ol - Undefined - ASS
Codec: ASS
AVC: No
Default: No
Forced: No
External: No

Subtitle
Title: Franأ§ais(France) - Undefined - ASS
Codec: ASS
AVC: No
Default: No
Forced: No
External: No

Subtitle
Title: Italiano - Undefined - ASS
Codec: ASS
AVC: No
Default: No
Forced: No
External: No

Subtitle
Title: Portuguأھs(Brasil) - Undefined - ASS
Codec: ASS
AVC: No
Default: No
Forced: No
External: No
```

**Comment by Valnjes** [comment]
Try LiteFin:
https://github.com/MoazSalem/litefin/

**Comment by katiekloss** [comment]
I strongly suspect this is a memory leak in the [JavascriptSubtitlesOctopus](https://github.com/jellyfin/JavascriptSubtitlesOctopus) component that jellyfin-web uses to render ASS subtitles, but I haven't located it yet.

You can reproduce the leak in a desktop browser by looking at the Javascript VM heap in DevTools while watching an affected video. The leak seems to be worse the larger the video is, which sounds like it's retaining the rendered subtitles for too long.
